### PR TITLE
CA Cert and Server Implementation 

### PR DESCRIPTION
**Pull request by kaalams**
This PR includes the implementation for CA-certificate generation, storing and retrieving these secrets from the GCP secret manager as well as unit tests for the proposed validation web hook.

### CODE REVIEW

File: prow/cmd/validation-webhook/main.go
Comment (chaodaiG) [inline prow/cmd/validation-webhook/main.go:50]: this is redundant, should be able to delete these two lines

File: prow/cmd/validation-webhook/main.go
Comment (chaodaiG) [inline prow/cmd/validation-webhook/main.go:57]: I'm not aware of the usage of `logrus.Print`, more commonly used are `.Info`, `.Warn`, `.Error` etc.

File: prow/cmd/validation-webhook/main.go
Comment (chaodaiG) [inline prow/cmd/validation-webhook/main.go:55]: actually I'm not entirely sure that this needs to be a function unless you have unit test for it

File: prow/cmd/validation-webhook/main.go
Comment (chaodaiG) [inline prow/cmd/validation-webhook/main.go:77]: this function is invoked when cert is not valid, which includes the case where the cert was expired, and in this case the GCP secret already exsits and `CreateSecret` would fail. Might want to do something like `client.GetSecret`, and Create only when get failed.

File: prow/cmd/validation-webhook/helpers.go
Comment (kaalams) [reply prow/cmd/validation-webhook/helpers.go:19]: yeah i got it from an engineering blog. I'll link it in the file

File: prow/cmd/validation-webhook/main.go
Comment (kaalams) [reply prow/cmd/validation-webhook/main.go:25]: typically wouldn't we like to have a standard/set projectid to store our gcpsecrets, seeing as the user has no control over this. This is for our cert generation not the users.

**Comment by fejta** [comment]
/uncc
Woo!

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [reply prow/cmd/validation-webhook/main.go:64]: It doesn't look like this function is actually used for any testing.
I was thinking of a test that ensure the certificates are working properly and that the `/validate` path is reachable. You can use the `httptest` package to spin up a server on an arbitrary free port: https://pkg.go.dev/net/http/httptest#NewUnstartedServer

File: prow/cmd/validation-webhook/main_test.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main_test.go:49]: After this line we don't actually call any non-test code and we don't check the values returned on this line so we aren't actually validating anything besides the fake client.
Rather than defining a `FakeClientInterface` here, define and use a `ClientInterface` in your main code and then pass the `FakeClient` rather than a real client to test functions like `getSecrets` and `createGCPSecret`.

The same applies to `TestSecretsDontExist`.

File: prow/cmd/validation-webhook/main_test.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main_test.go:11]: Types defined in `*_test.go` files shouldn't typically be exported (capitalized).

File: prow/cmd/validation-webhook/main.go
Comment (kaalams) [reply prow/cmd/validation-webhook/main.go:46]: im unmarshalling the map and returning the cert and privkey if getting the secret succeeds, I could make it global i guess?

File: prow/cmd/validation-webhook/main.go
Comment (kaalams) [reply prow/cmd/validation-webhook/main.go:55]: not exactly sure what you mean here

File: prow/cmd/validation-webhook/main.go
Comment (kaalams) [reply prow/cmd/validation-webhook/main.go:40]: so i should get rid of getSecretsIfPresent function?

Comment (cjwagner) [reply prow/cmd/validation-webhook/main.go]: I do not see a function with that name.

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [reply prow/cmd/validation-webhook/main.go]: It looks like you fixed this.
However I have an additional nit: When passing a returned error up to a higher level in the call stack, it is common to wrap it with some context. So rather than just returning `err` return something like `fmt.Errorf("error unmarshalling CA cert secret data: %w", err)`. Without this context it can be difficult to determine what we were trying to do when we encountered the error.
This isn't always necessary depending on the specific case, but please read through this PR and add context to errors that are being passed through functions where it seems helpful.

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [reply prow/cmd/validation-webhook/main.go]: It could just be a local parameter of the function. We don't actually use the map after `getSecrets` so we don't need a reference to it outside this function.

File: prow/cmd/validation-webhook/main.go
Comment (kaalams) [reply prow/cmd/validation-webhook/main.go]: my bad i meant getSecrets()

File: prow/cmd/validation-webhook/helpers.go
Comment (kaalams) [reply prow/cmd/validation-webhook/helpers.go:78]: not much info in the docs https://pkg.go.dev/crypto/x509#Certificate:~:text=type%20Certificate%20%C2%B6,.ObjectIdentifier%0A%7D

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main.go:63]: You can get rid of the `dnsNames` slice and this loop, they aren't needed.
```suggestion
	dns := flagutil.NewStrings("validation-webhook-service", "validation-webhook-service.default", "validation-webhook-service.default.svc")
```

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main.go:75]: This doesn't do anything, it returns a log entry that isn't used.

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main.go:80]: This will cause a segfault in the normal case where `err == nil`. I think you mean for this to be 
```suggestion
	} else if err != nil {
```

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main.go:84]: nit: For simple error checks like this it is common to use this format since the `err` is not needed outside the conditional if it is `nil`.
```suggestion
	if err := isCertValid(cert); err != nil {
```

If the cert is invalid and we need to refresh it we should log an info level message with the details about the expiration from `err`. Right now the error message text is unused.

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main.go:100]: This ignores the message contents in `err` and returns a potentially inaccurate error. `GetSecretValue` could fail due to auth issues or network problems, we don't know that the error indicates the secret doesn't exist.

Also if you want to distinguish an error you shouldn't match against error text like you do on line 74. It is preferable to use type checking instead to see if the error is of a specific type.

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main.go:114]: You don't need to pass the full `flagutil.Strings`. This will be easier to test if you just pass the `[]string` provided by `dns.Strings()`.

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main.go:127]: This is marshalling, not unmarshalling.

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [reply prow/cmd/validation-webhook/main.go:130]: This has not been resolved.

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main.go:39]: This cannot be used directly. You need to replace the `*`s with the correct values. Consider using a format string.

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main.go:134]: The second parameter to this function should be the secret name.

File: prow/cmd/validation-webhook/main.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main.go:34]: This is not specific enough, this is a project scoped identifier. Consider something like `prowjob-webhook-ca-cert`.

File: prow/cmd/validation-webhook/main_test.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main_test.go:85]: Currently there is a bug where the secret is stored under the wrong name. Please improve this test to catch that.

File: prow/cmd/validation-webhook/main_test.go
Comment (cjwagner) [inline prow/cmd/validation-webhook/main_test.go:31]: I don't understand the use of dirname here. The secret store doesn't have a concept of directories and this value isn't being compared against the secretName in `GetSecretValue` or `AddSecretValue` to see if the secret exists.

File: prow/cmd/validation-webhook/main.go
Comment (kaalams) [reply prow/cmd/validation-webhook/main.go:75]: yeah im just letting the user know that it could not get secret data, before trying to create the secret

File: prow/cmd/validation-webhook/main.go
Comment (kaalams) [reply prow/cmd/validation-webhook/main.go:100]: type error interface {
	Error() [string](https://pkg.go.dev/builtin#string)
}
this is what the error interface provides

Comment (kaalams) [reply prow/cmd/validation-webhook/main.go:100]: what exactly do you mean?

File: prow/cmd/validation-webhook/helpers.go
Comment (mpherman2) [reply prow/cmd/validation-webhook/helpers.go:2]: On the other files as well

File: prow/cmd/validation-webhook/helpers.go
Comment (mpherman2) [inline prow/cmd/validation-webhook/helpers.go:73]: nit: This is only used in one place. Is the var necessary?

File: prow/cmd/validation-webhook/main.go
Comment (mpherman2) [reply prow/cmd/validation-webhook/main.go:75]: This seems to be intended behavior. If there is an error here we attempt to create the secret? If so, we do not want to spam the logs with errors that we expect to hit. Maybe make this debug or info with more information that we are attempting to create the secret.

**Review by mpherman2 — COMMENTED** [review]
Looking good. Left a few comments.

File: prow/cmd/webhook-server/main.go
Comment (kaalams) [reply prow/cmd/webhook-server/main.go:69]: noticed a possible segfault here if theres no error. it won't be able to access err.Error() and error out. Will fix that

File: prow/cmd/webhook-server/main.go
Comment (mpherman2) [inline prow/cmd/webhook-server/main.go:67]: I don't think you even need this boolean. This can be simplified if you

```
if err != nil && strings.Contains(err.Error(), gcpSecretError) {
     // Secret does not exist yet, Create Secret
} else if err != nil {
  // return error
}
...
```

File: prow/cmd/webhook-server/main.go
Comment (mpherman2) [inline prow/cmd/webhook-server/main.go:74]: nit: would probably be cleaner to combine these

**Review by mpherman2 — APPROVED** [review]
LGTM
/hold
holding in case @cjwagner wants to take another look before we merge.

File: prow/cmd/webhook-server/main.go
Comment (cjwagner) [inline prow/cmd/webhook-server/main.go:123]: This function could just call `client.CreateSecret` and then call `updateGCPSecret` to reduce redundancy.

File: prow/cmd/webhook-server/main.go
Comment (cjwagner) [inline prow/cmd/webhook-server/main.go:151]: This is still not fixed. This is marshalling, not unmarshalling.

**Comment by cjwagner** [comment]
/lgtm cancel
Due to listenandserveTLS needing to use file paths.

**Review by mpherman2 — APPROVED** [review]
/unhold

**Pull request merged**
